src/lib/operators/table_scan/sorted_segment_between_search.hpp

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+#include <functional>
+#include <thread>
+#include <type_traits>
+
+#include <boost/range.hpp>
+#include <boost/range/join.hpp>
+
+#include "all_type_variant.hpp"
+#include "constant_mappings.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+// Generic class which handles the actual scanning of a sorted segment
+template <typename IteratorType, typename SearchValueType>
+class SortedSegmentBetweenSearch {
+ public:
+  SortedSegmentBetweenSearch(IteratorType begin, IteratorType end, const OrderByMode& order_by, const bool nullable,
+                             const PredicateCondition& predicate_condition, const SearchValueType& left_value,
+                             const SearchValueType& right_value)
+      : _begin{begin},
+        _end{end},
+        _predicate_condition{predicate_condition},
+        _left_value{left_value},
+        _right_value{right_value},
+        _nullable{nullable},
+        _is_ascending{order_by == OrderByMode::Ascending || order_by == OrderByMode::AscendingNullsLast},
+        _is_nulls_first{order_by == OrderByMode::Ascending || order_by == OrderByMode::Descending} {}
+
+ private:
+  void _exponential_search_for_nulls(IteratorType it_first, IteratorType it_last) {
+    if (it_first == it_last) return;
+    // If no null values are present
+    if (_is_nulls_first && !it_first->is_null()) {
+      return;
+    }
+    if (!_is_nulls_first && !(it_last - 1)->is_null()) {
+      return;
+    }
+
+    using difference_type = typename std::iterator_traits<IteratorType>::difference_type;
+    difference_type size = std::distance(it_first, it_last), bound = 1;
+
+    if (_is_nulls_first) {
+      while (bound < size && (it_first + bound)->is_null()) {
+        bound *= 2;
+      }
+
+      auto end = it_first + std::min(bound, size);
+      _begin = std::lower_bound(it_first + (bound / 2), end, false,
+                                [](const auto& segment_position, const auto& _) { return segment_position.is_null(); });
+    } else {
+      while (bound < size && (it_first + (size - bound))->is_null()) {
+        bound *= 2;
+      }
+
+      auto start = it_first + (size - std::min(bound, size));
+      _end = std::lower_bound(start, it_first + (size - bound / 2), true,
+                              [](const auto& segment_position, const auto& _) { return !segment_position.is_null(); });
```

Comment:
Hmm … I think that my description of the `std::lower_bound` call is wrong and should instead be: 

``` diff
- … for which the following condition holds true: The value is not NULL.
+ … for which the following condition does not hold true: The value is not NULL. 
+ (i. e. the first element which is NULL)
```

Thus, it seems like your implementation is perfectly fine. However, I'll leave the review comment here so that you can still have a look at it and double-check. I hope that's okay for you and doesn't lead to more confusion on your side as well. 😉

Reply:
Thank you very much for your deep thoughts about lower_bound and upper_bound and our algorithm. I checked it three times now (Ascending, Descending, AscendingNullsLast) and in my opinion, there should be no faults. I will discuss this with my teammates again, though, because you never know :)